lib/src/main/java/io/ably/lib/transport/ConnectionManager.java

```diff
@@ -71,6 +71,11 @@ public class ConnectionManager implements ConnectListener {
     static ErrorInfo REASON_REFUSED = new ErrorInfo("Access refused", 401, 40100);
     static ErrorInfo REASON_TOO_BIG = new ErrorInfo("Connection closed; message too large", 400, 40000);
 
+    /**
+     * Flag that indicating that we entered termination state
+     */
+    private boolean terminating = false;
```

Comment:
Since, this is being accessed from multiple threads, I guess we should use 
[AtomicBoolean](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/AtomicBoolean.html)

Reply:
Yeah, I kind of agree, `terminating` is not being set at lot of places.